source/extensions/transport_sockets/s2a/config.cc

```diff
@@ -0,0 +1,119 @@
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace S2A {
+namespace {
+
+// Manage S2A singleton state via SingletonManager
+class S2ASharedState : public Singleton::Instance {
+public:
+  S2ASharedState() { grpc_s2a_shared_resource_dedicated_init(); }
+
+  ~S2ASharedState() override { grpc_s2a_shared_resource_dedicated_shutdown(); }
+
+private:
+  // There is blanket google-grpc initialization in MainCommonBase, but that
+  // doesn't cover unit tests. However, putting blanket coverage in ProcessWide
+  // causes background threaded memory allocation in all unit tests making it
+  // hard to measure memory. Thus we also initialize grpc using our idempotent
+  // wrapper-class in classes that need it. See
+  // https://github.com/envoyproxy/envoy/issues/8282 for details.
+#ifdef ENVOY_GOOGLE_GRPC
+  Grpc::GoogleGrpcContext google_grpc_context_;
+#endif
+};
+
+SINGLETON_MANAGER_REGISTRATION(s2a_shared_state);
+
+Network::TransportSocketFactoryPtr createTransportSocketFactoryHelper(
+    const Protobuf::Message& message, bool is_upstream,
+    Server::Configuration::TransportSocketFactoryContext& factory_ctxt) {
+  // A reference to this is held in the factory closure to keep the singleton
+  // instance alive.
+  auto s2a_shared_state = factory_ctxt.singletonManager().getTyped<S2ASharedState>(
+      SINGLETON_MANAGER_REGISTERED_NAME(s2a_shared_state),
+      [] { return std::make_shared<S2ASharedState>(); });
+  auto config =
+      MessageUtil::downcastAndValidate<const envoy::extensions::transport_sockets::s2a::v3alpha::S2AConfiguration&>(
+          message, factory_ctxt.messageValidationVisitor());
+
+  const std::string& s2a_address = config.s2a_address();
+  HandshakerFactory factory =
+      [s2a_address, is_upstream,
+       s2a_shared_state](Event::Dispatcher& dispatcher,
+                          const Network::Address::InstanceConstSharedPtr& local_address,
+                          const Network::Address::InstanceConstSharedPtr&) -> TsiHandshakerPtr {
+    ASSERT(local_address != nullptr);
+
+    grpc_s2a_credentials_options* options = grpc_s2a_credentials_options_create();
+    grpc_s2a_credentials_options_set_s2a_address(options, s2a_address.c_str());
+    const char* target_name = is_upstream ? "" : nullptr;
+    s2a::tsi::S2ATsiHandshakerOptions s2a_options{/* interested_parties= */nullptr, is_upstream, options, target_name};
+    // Specifying target name as empty since TSI won't take care of validating peer identity
+    // in this use case. The validation will be performed by TsiSocket with the validator.
+    // Set the max frame size to 16KB.
+    absl::StatusOr<tsi_handshaker*> tsi_handshaker_ptr = CreateS2ATsiHandshaker(s2a_options);
+
+    if (!tsi_handshaker_ptr.ok()) {
+      const std::string handshaker_name = is_upstream ? "client" : "server";
+      ENVOY_LOG_MISC(warn, "Cannot create S2A {} handshaker, status: {}", handshaker_name, tsi_handshaker_ptr.status());
```

Comment:
Can we make sure to add a test that hits this if statement? I want to make sure we gracefully handle failure if it comes from underneath gRPC TSI.